### ISSUE TITLE
Refactor the ODE tests to reduce repetition and make the examples usable for other things

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#18377]
Currently there are a lot of repetitive tests for ODEs in `test_ode.py`:
https://github.com/sympy/sympy/blob/f5413ff89530b14ffc9b7359717f8c7ad03d2c6c/sympy/solvers/tests/test_ode.py#L2839-L2847

We should make a function that does this so that those ODEs can be put neatly into data structures. For example given
```python
def test_odes_basic():
    eq = Eq(f(x).diff(x), 0)
    sol = Eq(f(x), C1)
    assert dsolve(eq, f(x)) == sol
    assert checkodesol(eq, sol)

    eq = Eq(f(x).diff(x), 1)
    sol = Eq(f(x), x + C1)
    assert dsolve(eq, f(x)) == sol
    assert checkodesol(eq, sol)
```
We could rewrite that as
```python
_odesol_basic = {
    'basic1': {
        'eq': Eq(f(x).diff(x), 0),
        'sol': Eq(f(x), C1),
    }
    'basic2': {
        'eq': Eq(f(x).diff(x), 1),
        'sol': Eq(f(x), x + C1),
    }
}

def test_ode_basic():
    _test_all_odesol(_odesol_basic)
```
The idea is that the `test_all` function can take care of ensuring that the equation and solution are fully tested in the normal way. For some cases there could be other parameters like `hint` or `func` (if the function is not `f(x)`). I guess some ODEs will have different solutions for different hints so they could be listed in some way. There will need to be some way to list XFAILs as well.

This can improve the tests:
1. We can make sure that all examples are fully tested properly
2. Some tests can be made faster e.g. it is probably not necessary to call `dsolve` twice with different hints.
3. Sometimes `dsolve` is tested with a particular hint but calling `dsolve` without the hint would invoke a different solver and we don't know that it would actually work.

Apart from making the test code more systematic the advantage of splitting out the different test examples into data structures is that we can use them for other things besides just running the normal test suite:
1. We can make a mega-test that tests every ODE against every possible hint. This would be too slow for the normal test suite but I am sure that many bugs would immediately be found.
2. We can use the examples for benchmarking to measure the performance of dsolve.

**Comment by oscarbenjamin** [comment]
Note that this issue can be done over many pull requests. It does not all need to be done in one go.

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin can this be included in the gsoc idea of refactoring ode module?

**Comment by oscarbenjamin** [comment]
Yes, ideally the tests for the solvers that are moving to single.py should also be moved to `test_single.py`

**Comment by Mohitbalwani26** [comment]
So for example the test related to Nth algebraic should be moved to single.py right?

**Comment by oscarbenjamin** [comment]
Yes.

**Comment by Mohitbalwani26** [comment]
So i will open a PR which will shift `Nth algebraic` , `FirstLinear` and `AlmostLinear` tests in `test_single.py` once #18403 is merged

**Comment by Mohitbalwani26** [comment]
oscarbenjamin for example i am moving nth_algebraic test in test_single.py so should i just ove test_nth_algebraic () function only or i have to move the test_issue() which are for nth_Alegbraic also?

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin  for example i am moving `nth_algebraic` test in test_single.py so should i just move `test_nth_algebraic ()`function only or i have to move the test_issue() which are for nth_Alegbraic also?

**Comment by oscarbenjamin** [comment]
Yes all the tests related to `nth_algebraic` and the other solvers in single.py

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin I looked at `test_ode.py` for every solver there is a separate function for its tests. so the idea discussed in OP should be implemented for particular solver? 
can you please elaborate more with example of let's say def `test_nth_linear_constant_coeff_undetermined_coefficients():`

**Comment by sylee957** [comment]
Are the refractories useful for tests?
For some cases, it can be useful, but I was rather annoyed by some complicated logic introduced in tests.
And I see that the overall lines have just increased for most of the PRs.

**Comment by oscarbenjamin** [comment]
The logic can probably be simplified. What is the problem with the traceback?

Ideally we would do this using pytest's parametrised tests feature (#15497)

The refactor does not necessarily reduce the number of lines because we end up with effectively a lot of whitespace around the tests e.g. compare these:
https://github.com/sympy/sympy/blob/94fb720696f5f5d12bad8bc813699fd696afd2fb/sympy/solvers/ode/tests/test_ode.py#L434-L442
https://github.com/sympy/sympy/blob/94fb720696f5f5d12bad8bc813699fd696afd2fb/sympy/solvers/ode/tests/test_single.py#L443-L451

I think it's easier to see the actual ODE and solution in the refactored tests. It also means we can be more explicit about what is or is not tested for each example. From a quick skim of the test file it's not immediately obvious that the tests above don't actually test that dsolve can find the solution.

The main advantage though is that we can use each test example for extra tests or benchmarks that aren't part of the normal test suite. Some of the ODE solvers are not tested with many examples but this makes it possible to test every example against every ODE and some bugs have already been found this way.

**Comment by sylee957** [comment]
Okay, I didn't notice before that it uses custom error messages.

**Comment by oscarbenjamin** [comment]
There has been a lot of progress on this thanks to @Mohitbalwani26.

There are still examples of ODEs and solutions in there that would ideally be reorganised into the same format in `test_single.py` though: group the examples by hint and then flag which parts of the example doesn't work (e.g. if XFAIL doesn't work or whatever).

Also the systems of ODEs examples should be moved from `test_ode.py` to `test_systems.py` (even if they are not otherwise refactored in any way).

Another problem with the ODE tests is that there are some tests that are not marked as slow that are actually very slow:
```
________________________________ slowest tests _________________________________
sympy/solvers/ode/tests/test_systems.py::test_component_division - Took 10.031 seconds
sympy/stats/tests/test_joint_rv.py::test_GeneralizedMultivariateLogGammaDistribution - Took 11.075 seconds
sympy/utilities/tests/test_wester.py::test_W19 - Took 11.092 seconds
sympy/stats/tests/test_finite_rv.py::test_bernoulli - Took 11.240 seconds
sympy/stats/tests/test_compound_rv.py::test_normal_CompoundDist - Took 11.610 seconds
sympy/solvers/ode/tests/test_single.py::test_Liouville_ODE - Took 12.471 seconds
sympy/stats/tests/test_finite_rv.py::test_ideal_soliton - Took 12.752 seconds
sympy/solvers/tests/test_solveset.py::test_linsolve_large_sparse - Took 12.763 seconds
sympy/solvers/ode/tests/test_ode.py::test_1st_homogeneous_coeff_ode_check7 - Took 13.400 seconds
sympy/solvers/ode/tests/test_single.py::test_nth_order_linear_euler_eq_nonhomogeneous_variation_of_parameters - Took 13.576 seconds
sympy/solvers/ode/tests/test_single.py::test_nth_linear_constant_coeff_undetermined_coefficients - Took 13.831 seconds
sympy/stats/tests/test_continuous_rv.py::test_exponential - Took 13.855 seconds
sympy/solvers/ode/tests/test_single.py::test_nth_order_linear_euler_eq_nonhomogeneous_undetermined_coefficients - Took 14.264 seconds
sympy/stats/tests/test_joint_rv.py::test_JointPSpace_marginal_distribution - Took 14.405 seconds
sympy/solvers/ode/tests/test_systems.py::test_higher_order_to_first_order - Took 16.255 seconds
sympy/solvers/tests/test_solvers.py::test_solve_args - Took 16.727 seconds
sympy/solvers/ode/tests/test_ode.py::test_exact_enhancement - Took 20.155 seconds
sympy/solvers/ode/tests/test_single.py::test_nth_linear_constant_coeff_var_of_parameters - Took 20.808 seconds
sympy/utilities/tests/test_wester.py::test_W9 - Took 21.931 seconds
sympy/solvers/ode/tests/test_single.py::test_1st_exact - Took 23.081 seconds
sympy/solvers/ode/tests/test_single.py::test_nth_linear_constant_coeff_homogeneous - Took 23.855 seconds
sympy/solvers/ode/tests/test_single.py::test_2nd_linear_bessel_equation - Took 25.213 seconds
sympy/solvers/ode/tests/test_ode.py::test_solve_ics - Took 25.732 seconds
sympy/solvers/ode/tests/test_single.py::test_nth_algebraic - Took 33.521 seconds
sympy/solvers/ode/tests/test_single.py::test_separable_reduced - Took 34.327 seconds
sympy/solvers/ode/tests/test_single.py::test_2nd_2F1_hypergeometric - Took 58.450 seconds
sympy/solvers/ode/tests/test_single.py::test_lie_group - Took 65.585 seconds
sympy/solvers/ode/tests/test_single.py::test_factorable - Took 80.202 seconds
sympy/solvers/ode/tests/test_single.py::test_separable - Took 106.990 seconds
```
We should at least mark these as slow so that they don't have to run as part of the standard test suite every time.

Ideally though we would use the slow flag on individual examples that are slow to separate the slow and the fast examples. What we want is for the not-slow tests to be the fastest tests that give good line coverage of the ODE module. Everything else should be marked as slow. Many of the slower tests are very specific regression tests that don't necessarily add anything to general coverage of the ODE code.

**Comment by oscarbenjamin** [comment]
> most of the remaining examples were added because of bug in checkodesol or some other fix so how should they be refactored?

Could you give an example?

> Also many of them have parameters like `{eta(x, f(x)): -a/b, xi(x, f(x)): 1}` so these should be added as keys in `test_single.py`?

Non-standard tests like that can be separate from the main body of tests.

**Comment by Mohitbalwani26** [comment]
> Could you give an example?
> 
something like this there are few test functions
```
def test_issue_4838():
    # Issue #15999
    eq = f(x).diff(x) - C1*f(x)
    sol = Eq(f(x), C2*exp(C1*x))
    assert dsolve(eq, f(x)) == sol
    assert checkodesol(eq, sol, order=1, solve_for_func=False) == (True, 0)
```

> Non-standard tests like that can be separate from the main body of tests.

so how they should be copied as it is or some other way? to be precise most of them are tests related to heuristics.

**Comment by oscarbenjamin** [comment]
> so how they should be copied as it is or some other way? to be precise most of them are tests related to heuristics.

We could probably make a separate file called `test_lie_group.py` and move those tests there.

> something like this there are few test functions

Are you querying the use of `solve_for_func`? It isn't necessary in this example.

**Comment by Mohitbalwani26** [comment]
> We could probably make a separate file called `test_lie_group.py` and move those tests there.

so what about the examples of lie group which are currently in `test_single.py` they should also be moved or not?

> > something like this there are few test functions
> 
> Are you querying the use of `solve_for_func`? It isn't necessary in this example.

No, I was asking that the tests which were added as a bug fix or some changes other than hints what should be done with that?
like the example which I stated above was added due to issue #4838 and it doesn't specifically target one hint to categorise its examples.

**Comment by oscarbenjamin** [comment]
> so what about the examples of lie group which are currently in `test_single.py` they should also be moved or not?

I think that the ones that just use a plain `dsolve` call can go in with the main tests. The tests that use specific infinitesimal or heuristic arguments should be in `test_lie_group.py`.

> No, I was asking that the tests which were added as a bug fix or some changes other than hints what should be done with that?
> like the example which I stated above was added due to issue #4838 and it doesn't specifically target one hint to categorise its examples.

We could just have a miscellaneous examples section.
